Make the HTTP.sys accept loop cheaper

davidfowl opened the pull request:
- Allocate a single AsyncAcceptContext per loop instead of one per request.
- The AsyncAcceptContext now implements IValueTaskSource instead of
Task and doesn't allocate a Task per request.
- Use a preallocated overlapped since we don't have overlapping calls to accept for a single AsyncAcceptContext.
- Remove some dead code from SafeNativeOverlapped
- Move the NativeOverlapped* to the AsyncAcceptContext

Contributes to #22022

davidfowl committed:
Make the HTTP.sys accept loop cheaper
- Allocate a single AsyncAcceptContext per loop instead of one per request.
- The AsyncAcceptContext now implements IValueTaskSource instead of
Task and doesn't allocate a Task per request.
- Use a preallocated overlapped since we don't have overlapping calls to accept for a single AsyncAcceptContext.
- Remove some dead code from SafeNativeOverlapped

davidfowl commented on src/Servers/HttpSys/src/AsyncAcceptContext.cs:
```diff
@@ -2,121 +2,132 @@
+            if (statusCode != UnsafeNclNativeMethods.ErrorCodes.ERROR_SUCCESS &&
+                statusCode != UnsafeNclNativeMethods.ErrorCodes.ERROR_IO_PENDING)
+            {
+                // some other bad error, possible(?) return values are:
```
No disposal is done here since the caller will end up calling AcceptAsync again and that will end up clearing the previously allocated request context (same goes for all other places that used to call dispose).

davidfowl committed:
Clear the nativeRequestContext on dispose

gfoidl commented on src/Servers/HttpSys/src/AsyncAcceptContext.cs:
```diff
@@ -2,121 +2,132 @@
+        private readonly PreAllocatedOverlapped _preallocatedOverlapped;
+
+        // mutable struct; do not make this readonly
+        private ManualResetValueTaskSourceCore<RequestContext> _tcs = new()
```
Nit about the naming of `_tcs`.

It matches well for TaskCompletionSource, but for the ManualResetValueTaskSourceCore...I'd name it `_mrvts` or maybe `_mrvtsc` or `_valueTaskSource` or ...

jkotalik commented on src/Servers/HttpSys/src/MessagePump.cs:
```diff
@@ -176,14 +176,17 @@ internal void SetShutdownSignal()
                 try
                 {
-                    requestContext = await Listener.AcceptAsync().SupressContext();
+                    requestContext = await Listener.AcceptAsync(acceptContext);
```
Confused why this was removed? Is SuppressContext used elsewhere?

davidfowl replied:
See https://github.com/dotnet/aspnetcore/pull/28308

davidfowl committed:
PR feedback
- Move NativeOverlapped* to the AsyncAcceptContext. This removes the SafeNativeOverlapped allocation as the lifetime is managed by the AsyncAcceptContext
- Rename tcs

davidfowl committed:
Merge branch 'master' into davidfowl/accept-loop

davidfowl commented:
@Tratcher https://github.com/dotnet/aspnetcore/pull/28347 depends on this PR so is that approve transitive?

Tratcher commented:
No, just the diff. I'll take a look at this one soon.

davidfowl commented on src/Servers/HttpSys/src/AsyncAcceptContext.cs:
```diff
@@ -2,121 +2,133 @@
-                        }
-                        catch (Exception)
-                        {
-                            server.SendError(asyncContext._nativeRequestContext.RequestId, StatusCodes.Status400BadRequest);
```
I was wondering if you'd catch it. I wanted to see if we thought it was important. I'll move it to the MessagePump is the accept fails.

davidfowl replied:
Exceptions should bubble to the caller

Tratcher replied:
> I'll move it to the MessagePump is the accept fails.

How? The message pump doesn't have access to the request if the Accept failed part way through.



> Exceptions should bubble to the caller

It can, but the error response still needs to be sent in order to prevent a native memory leak.

davidfowl replied:
I might be missing something but it seems the only thing passed from the request was the ID. I would make a new exception type that stored the ID and use that to write the response.

davidfowl committed:
Send the error response before settings the result
